### PR TITLE
Support custom metadata log service implementation

### PR DESCRIPTION
**Pull request by seankao-az**
### Description
1. Use reflection for `FlintMetadataLogServiceBuilder` to support custom implementation of `FlintMetadataLogService`
2. Read `SparkConf` for such builder
3. Make `FlintMetadataLogService` an abstract class with a constructor using `SparkConf`
4. Move interfaces to flint-commons
    1. Moved files: `FlintMetadataLog`, `FlintMetadataLogEntry`, `FlintMetadataLogService`, `OptimisticTransaction`
    2. Remove unused `failLogEntry` from `FlintMetadataLogEntry` (no access to org.opensearch.index namespace). Can add back in next PR when storage is abstracted away from the common `FlintMetadataLogEntry`

### Issues Resolved
* #371 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java [hunk 1]
```diff
@@ -5,14 +5,29 @@
...
+  public static FlintMetadataLogService build(FlintOptions options, SparkConf sparkConf) {
+    String className = options.getCustomFlintMetadataLogServiceClass();
+    if (className.isEmpty()) {
+      return new FlintOpenSearchMetadataLogService(options);
+    }
+
+    // Attempts to instantiate Flint metadata log service with sparkConf using reflection
+    try {
+      Class<?> flintMetadataLogServiceClass = Class.forName(className);
+      Constructor<?> constructor = flintMetadataLogServiceClass.getConstructor(SparkConf.class);
```
Comment (penghuo) [inline flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java:27]: it required FlintMetadataLogService constructor has SparkConf as parameter. could we add doc on FlintMetadataLogService to clarify on it?

Comment (seankao-az) [reply flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java:27]: good point. Enforcing the constructor by making it a abstract class instead. what do you think?

Comment (penghuo) [reply flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java:27]: it is ok for now, add doc should be fine.

**Review by penghuo — APPROVED** [review]
thx!

File: flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java (hunk 1 above)
Comment (seankao-az) [reply flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java:31]: just realized there really is no way to enforce implementation class to have a specific signature for constructor (my attempt doesn't work). Reverting it and add doc instead.

File: flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java [hunk 2]
```diff
@@ -5,14 +5,33 @@
...
 import org.opensearch.flint.core.storage.FlintOpenSearchMetadataLogService;
 
 /**
  * {@link FlintMetadataLogService} builder.
+ * <p>
+ * Custom implementations of {@link FlintMetadataLogService} are expected to provide a public
+ * constructor with the signature {@code public MyCustomService(SparkConf sparkConf)} to be
+ * instantiated by this builder.
  */
 public class FlintMetadataLogServiceBuilder {
```
Comment (noCharger) [inline flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java:21]: should FlintMetadataLogServiceBuilder moved to common as well?

File: flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java (hunk 1 above)
Comment (noCharger) [reply flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java:31]: Is this change reverted? it seems still use `SparkConf `

File: flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java (hunk 2 above)
Comment (seankao-az) [reply flint-core/src/main/scala/org/opensearch/flint/core/metadata/log/FlintMetadataLogServiceBuilder.java:21]: no this needs access to `FlintOpenSearchMetadataLogService`

**Pull request merged**